File: python/mirage/core/chroma/stat.py
from mirage.cache.index import IndexCacheStore
from mirage.core.chroma.path import resolve_path
from mirage.types import FileStat, FileType, PathSpec
from mirage.utils.errors import enoent


async def stat_light(accessor, path: PathSpec,
                     index: IndexCacheStore) -> FileStat:
    return await stat(accessor, path, index)


async def stat(accessor, path: PathSpec, index: IndexCacheStore) -> FileStat:
    resolved = await resolve_path(accessor, path, index)
    if resolved.is_dir:
        return FileStat(
            name=stat_name(resolved.virtual_key, resolved.mount_prefix),
            type=FileType.DIRECTORY,
            extra={"children_count": 0},
        )
    if resolved.entry is None:
        raise enoent(path)
    return FileStat(
        name=resolved.entry.name,
        type=FileType.TEXT,
        size=resolved.entry.size,
        modified=resolved.entry.extra.get("updated_at"),
        fingerprint=None,
        revision=None,
        extra=dict(resolved.entry.extra),
    )


def stat_name(virtual_key: str, mount_prefix: str) -> str:
    root = mount_prefix.rstrip("/") or "/"
    if virtual_key == root:
        return "/"
    return virtual_key.rstrip("/").rsplit("/", 1)[-1]

File: typescript/packages/core/src/core/chroma/stat.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { ChromaAccessor } from '../../accessor/chroma.ts'
import type { IndexCacheStore } from '../../cache/index/store.ts'
import { FileStat, FileType, PathSpec } from '../../types.ts'
import { resolvePath } from './path.ts'
import { enoent } from '../../utils/errors.ts'

export function statLight(
  accessor: ChromaAccessor,
  path: PathSpec | string,
  index?: IndexCacheStore,
): Promise<FileStat> {
  return stat(accessor, path, index)
}

export async function stat(
  accessor: ChromaAccessor,
  path: PathSpec | string,
  index?: IndexCacheStore,
): Promise<FileStat> {
  const spec = typeof path === 'string' ? PathSpec.fromStrPath(path) : path
  const resolved = await resolvePath(accessor, spec, index)
  if (resolved.isDir) {
    return new FileStat({
      name: statName(resolved.virtualKey, resolved.mountPrefix),
      type: FileType.DIRECTORY,
      extra: { children_count: 0 },
    })
  }
  if (resolved.entry === null) throw enoent(spec.original)
  const updatedAt = resolved.entry.extra.updated_at
  return new FileStat({
    name: resolved.entry.name,
    type: FileType.TEXT,
    size: resolved.entry.size,
    modified: typeof updatedAt === 'string' ? updatedAt : null,
    extra: { ...resolved.entry.extra },
  })
}

export function statName(virtualKey: string, mountPrefix: string): string {
  const root = mountPrefix.replace(/\/+$/, '') !== '' ? mountPrefix.replace(/\/+$/, '') : '/'
  if (virtualKey === root) return '/'
  const stripped = virtualKey.replace(/\/+$/, '')
  return stripped.split('/').pop() ?? '/'
}
